### PR TITLE
PC regression in Jax

### LINKED ISSUE #8: Implement pc regression with jax
Original source: 

https://github.com/theislab/scib/blob/main/scib/metrics/pcr.py

See the manuscript for a description (search principal component regression):

https://www.nature.com/articles/s41592-021-01336-8#Sec11

### PR DESCRIPTION
**Pull request by martinkim0**
Closes #8.

### CODE REVIEW

**Comment by martinkim0** [comment]
I'll share a notebook later, but the implementation output matches the original `scib` package. It differs a bit when covariates are categorical because of the way `pd.get_dummies` handles one-hot labels, might need to look a bit more into this.

**Comment by adamgayoso** [comment]
> I'll share a notebook later, but the implementation output matches the original scib package. It differs a bit when covariates are categorical because of the way pd.get_dummies handles one-hot labels, might need to look a bit more into this.

This seems important, could it be an extra intercept term somewhere?

File: src/scib_metrics/_pcr.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+import jax
+import jax.numpy as jnp
+
+from scib_metrics.utils import one_hot, pca
+
+from ._types import NdArray
+
+
+def pcr(
```
Comment (adamgayoso) [inline src/scib_metrics/_pcr.py:11]: let's use the full name here

File: src/scib_metrics/_pcr.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    Computes the overall variance contribution given a covariate according to the following formula:
+
+    # TODO: Add formula
+
+    Parameters
+    ----------
+    X
+        Array of shape (n_samples, n_features).
+    batch
```
Comment (adamgayoso) [inline src/scib_metrics/_pcr.py:27]:
```suggestion
    covariate
```

File: tests/test_basic.py [hunk 3]
```diff
@@ -64,3 +64,29 @@ def test_kmeans():
...
+    assert pca.svd.s.shape == (max_components,)
+    assert pca.svd.v.shape == (max_components, X.shape[1])
+
+
+def test_pcr():
+    X, _, batch = dummy_x_labels_batch()
+    max_components = min(X.shape)
+    n_components = max_components - 1
+
+    _ = scib_metrics.pcr(X, batch, categorical=True, n_components=n_components)
```
Comment (adamgayoso) [inline tests/test_basic.py:91]: let's add scib as a pip package for testing and test against their implementation

**Review by adamgayoso — CHANGES_REQUESTED** [review]
cool!

**Comment by martinkim0** [comment]
@adamgayoso Looks like `pd.get_dummies` doesn't actually use one-hot encodings. I'll change the code to use it instead.

File: tests/utils/test_pca.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        if n_components < max_components:
+            sklearn_pca = PCA(n_components=n_components, svd_solver="arpack")
+            pca_true = sklearn_pca.fit(X)
+            assert jnp.allclose(sklearn_pca.fit_transform(X), pca.coordinates)
+            assert jnp.allclose(pca_true.components_, pca.components)
+            assert jnp.allclose(pca_true.explained_variance_, pca.variance)
+            assert jnp.allclose(pca_true.explained_variance_ratio_, pca.variance_ratio)
+
+    for n_obs in [10, 100, 1000]:
+        for n_vars in [10, 100, 1000]:
```
Comment (adamgayoso) [inline tests/utils/test_pca.py:43]: https://docs.pytest.org/en/6.2.x/parametrize.html

File: src/scib_metrics/utils/_utils.py [hunk 5]
```diff
@@ -1,10 +1,17 @@
...
+import jax
 import jax.numpy as jnp
+import numpy as np
 
 from .._types import NdArray
 
 
+def get_ndarray(x: jnp.ndarray) -> np.ndarray:
+    """Convert Jax device array to Numpy."""
+    return np.asarray(jax.device_get(x))
```
Comment (adamgayoso) [inline src/scib_metrics/utils/_utils.py:12]: maybe try np.array so it explicitly makes a copy

File: src/scib_metrics/_pcr.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+from scib_metrics.utils import one_hot, pca
+
+from ._types import NdArray
+
+
+def principal_component_regression(
+    X: NdArray,
+    covariate: NdArray,
+    categorical: Optional[bool] = False,
```
Comment (justjhong) [inline src/scib_metrics/_pcr.py:14]: nit: Not optional if default is False (it's just an alias for Union[bool, None])

File: src/scib_metrics/utils/_pca.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+import jax
+import jax.numpy as jnp
+
+from .._types import NdArray
+from ._utils import get_ndarray
+
+
+@dataclass
+class _SVD:
```
Comment (justjhong) [inline src/scib_metrics/utils/_pca.py:12]: just to avoid conflicting with SVD running the algorithm itself, I would call this SVDOutput or SVDResult

File: src/scib_metrics/utils/_pca.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        Array of shape (n_components, n_features) containing the right singular vectors.
+    """
+
+    u: NdArray
+    s: NdArray
+    v: NdArray
+
+
+@dataclass
+class _PCA:
```
Comment (justjhong) [inline src/scib_metrics/utils/_pca.py:31]: same comment here

File: src/scib_metrics/utils/_utils.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+from .._types import NdArray
+
+
+def get_ndarray(x: jnp.ndarray) -> np.ndarray:
+    """Convert Jax device array to Numpy."""
+    return np.array(jax.device_get(x))
+
+
+def one_hot(y: NdArray, n_classes: Optional[int] = None) -> NdArray:
```
Comment (justjhong) [inline src/scib_metrics/utils/_utils.py:15]: have you looked at https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.one_hot.html?

File: tests/test_pcr.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+import scib_metrics
+from scib_metrics.utils import get_ndarray
+
+from .utils.sampling import categorical_sample, normal_sample, poisson_sample
+
+PCR_PARAMS = list(product([10, 100, 1000], [10, 100, 1000], [True, False]))
+
+
+@pytest.mark.parametrize("n_obs, n_vars, categorical", PCR_PARAMS)
```
Comment (justjhong) [inline tests/test_pcr.py:15]: legendary

File: src/scib_metrics/utils/_utils.py (hunk 9 above)
Comment (martinkim0) [reply src/scib_metrics/utils/_utils.py:15]: Not sure how I missed that, thanks!

**Comment by martinkim0** [comment]
Anyone know how to solve the issue with `pytest` not recognizing the `tests` directory as a package? Issue is solved when I run `python -m pytest` but not when just running `pytest`.

**Comment by adamgayoso** [comment]
Has something to do with my last commit and this:

https://docs.pytest.org/en/7.1.x/explanation/pythonpath.html

File: src/scib_metrics/_pcr.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        Principal component regression using the first n_components principal components.
+    """
+    if len(X.shape) != 2:
+        raise ValueError("Dimension mismatch: X must be 2-dimensional.")
+    if X.shape[0] != covariate.shape[0]:
+        raise ValueError("Dimension mismatch: X and batch must have the same number of samples.")
+
+    covariate = one_hot(covariate) if categorical else covariate.reshape((covariate.shape[0], 1))
+
+    pca_results = pca(X, n_components=n_components)
```
Comment (adamgayoso) [inline src/scib_metrics/_pcr.py:46]: realizing that if X is already low-dim we won't see much of a speedup here, so we could opt to use sklearn pca if it helps reproduce their metric

**Comment by martinkim0** [comment]
Hmm not sure why tests are not passing since it seems to be working fine locally.

**Comment by adamgayoso** [comment]
@justjhong can you try this test locally on a linux machine to see if it passes for you?

**Comment by justjhong** [comment]
they pass for me too. However, I had this issue before where the random seed might be slightly different or the numpy version is slightly different on the test node. With these really small values for large vectors, you should just make the allclose check more lenient like here: https://github.com/YosefLab/scib-metrics/pull/20/files#diff-29093b485f46bc37dda5e0854cfc54e289f24e206207afb81dff1c527cb001a7R42

**Comment by martinkim0** [comment]
Local vs. workflow testing differences seem to come down to whether GPU is used or not, confirmed by comparing outputs two fresh conda environments. Interestingly, the values returned by the Jax implementation does not differ between the environments, but the ones from the original `scib` implementation do.

File: src/scib_metrics/_pcr.py (hunk 11 above)
Comment (martinkim0) [reply src/scib_metrics/_pcr.py:43]: Resolving per comment below.

**Review by adamgayoso — APPROVED** [review]
